### PR TITLE
Fix #17125 - Invalid DEX section size calculation issue ##bin

### LINKED ISSUE #17125: Huge section virtual size in bin_dex
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Fedora 31 x86_64
| Architecture/bits of the file (mandatory)            | DEX
| r2 -v full output, **not truncated** (mandatory)     | radare2 4.5.0-git 27197 @ linux-x86-64 git.4.4.0-314-gcfd23f5e3 commit: cfd23f5e3c7b0f189bd3f2c3c4b83ba76a248ef3 build: 2020-06-19__19:54:30

### Expected behavior
I don't expect such a big `constpool` section virtual size. This *huge* virtual size cause the analysis to never end, because the entire space is scanned.

### Actual behavior
```
[0x00000003]> iS
[Sections]

nth paddr                     size vaddr                    vsize perm name
―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
0   0x00000000                0x70 0x00000000                0x70 -r-- header
1   0x00000070  0xffffffffffffff93 0x00000070  0xffffffffffffff93 -r-- constpool
2   0x00000003          0xe9caa270 0x00000003          0xe9caa270 -r-x code
3   0x000022e7               0xbee 0x000022e7               0xbee -r-- data
4   0x00000000              0x2ed5 0x00000000              0x2ed5 -r-- file
[0x00000003]> o
 3 * r-x 0x00002ed5 test/bins/fuzzed/clusterfuzz-testcase-minimized-ia_fuzz-5643065568526336.dms
[0x00000003]> om
 2 fd: 3 +0x00000003 0x00000003 - 0xe9caa272 r-x fmap.code
 1 fd: 3 +0x00000000 0x00000000 - 0x00002ed4 r-- fmap.file
```

### Steps to reproduce the behavior 
`radare2 -A test/bins/fuzzed/clusterfuzz-testcase-minimized-ia_fuzz-5643065568526336.dms`

### Additional Logs, screenshots, source-code,  configuration dump, ...
An initial attempt was done in https://github.com/radareorg/radare2/pull/17110, but it did not work. As I am not very familiar with the format, I am creating this bug.

### PR DESCRIPTION
**Pull request by trufae**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

radare2 -A test/bins/fuzzed/clusterfuzz-testcase-minimized-ia_fuzz-5643065568526336.dms

**Closing issues**

#17125

### CODE REVIEW

File: libr/bin/p/bin_dex.c [hunk 1]
```diff
@@ -1887,117 +1884,136 @@ static char *getname(RBinFile *bf, int type, int idx, bool sd) {
...
+
+static void add_segment(RList *ret, const char *name, Section s, int perm) {
+	RBinSection *bs = add_section (ret, name, s, perm, NULL);
+	if (bs) {
+		bs->is_segment = true;
+		bs->add = true;
+	}
+}
+
+static bool dim(const char *name, Section *pre, Section *cur, Section *nex, Section *all) {
```
Comment (ret2libc) [inline libr/bin/p/bin_dex.c:1919]: what does this name `dim` mean?

File: libr/bin/p/bin_dex.c [hunk 2]
```diff
@@ -1887,117 +1884,136 @@ static char *getname(RBinFile *bf, int type, int idx, bool sd) {
...
+	r_return_val_if_fail (cur && all, false);
+	if (pre) {
+		if (cur->addr < (pre->addr + pre->size)) {
```
Comment (ret2libc) [inline libr/bin/p/bin_dex.c:1922]: `pre && cur->addr < pre->addr + pre->size`

File: libr/bin/p/bin_dex.c (hunk 1 above)
Comment (trufae) [reply libr/bin/p/bin_dex.c:1919]: i took DIM from qbasic https://www.qb64.org/wiki/DIM it's kind of a cast, so it bounds the value in a variable to some restrictions of type and size. it's probably not the best name, any proposal?

Comment (ret2libc) [reply libr/bin/p/bin_dex.c:1919]:
```suggestion
static bool is_valid_section(const char *name, Section *pre, Section *cur, Section *nex, Section *whole) {
```

File: libr/bin/p/bin_dex.c [hunk 3]
```diff
@@ -2104,6 +2115,11 @@ static int cmp_path(const void *a, const void *b) {
 	return strcmp ((const char*)a, (const char*)b);
 }
 
+static bool is_classes_dex(const char *filename) {
+	return r_str_startswith (filename, "classes") \
```
Comment (ret2libc) [inline libr/bin/p/bin_dex.c:2119]: IMHO this splitting just makes reading more complicate. Without it the line wouldn't really be that long.

Comment (trufae) [reply libr/bin/p/bin_dex.c:2119]: there's no need to mess a conditional with two function calls and handcrafted strings, when the function name can clearly specify what is going to check in there. it is more readable imho

File: libr/bin/p/bin_dex.c (hunk 1 above)
Comment (trufae) [reply libr/bin/p/bin_dex.c:1919]: is not really just checkinf if its a valid section.its also trimming it's size and warning when something is wrong

Comment (ret2libc) [reply libr/bin/p/bin_dex.c:1919]: `validate_section` or `is_valid_section` or something like that :D `dim` is really hard name to understand. it remembers me about visual basic though :P

Comment (trufae) [reply libr/bin/p/bin_dex.c]: And you don't like those memories? :P 

changed

**Pull request merged**